cpp/recognition/planar/HopcroftTarjan.cpp

```diff
@@ -0,0 +1,303 @@
+#include <recognition/planar/HopcroftTarjan.hpp>
+#include <networkit/linkprediction/NeighborhoodUtility.hpp>
+
+namespace Koala {
+
+    HopcroftTarjan::HopcroftTarjan(NetworKit::Graph &graph, bool embedding): IPlanarGraphRecognition(graph, embedding){ }
+
+    void HopcroftTarjan::dfs(NetworKit::Graph &g,
+                            NetworKit::node v, 
+                            std::vector<bool> & reached
+                            )
+    {
+        reached[v] = true;
+        for(auto u: g.neighborRange(v)){
+            if(!reached[u]){
+                dfs(g, u, reached);
+            }
+        }
+    }
+
+    void HopcroftTarjan::dfs_in_make_biconnected_graph(NetworKit::Graph &g, 
+                                                    NetworKit::node v, 
+                                                    int & dfs_count, 
+                                                    std::vector<bool> & reached,
+                                                    std::vector<NetworKit::node> & dfsnum,
+                                                    std::vector<NetworKit::node> &lowpt, 
+                                                    std::vector<NetworKit::node> &parent)
+    {
+        NetworKit::node w = INT_MAX;
+        dfsnum[v] = dfs_count++;
+        lowpt[v] = dfsnum[v];
+        reached[v] = true;
+        std::vector<int> neighbors;
+        for(auto u: g.neighborRange(v)) {
+            neighbors.push_back(u);
+        }
+        for(auto u: neighbors){
+            if(!reached[u]){
+                if(w == INT_MAX){
+                    w = u;
+                }
+                parent[u] = v;
+                dfs_in_make_biconnected_graph(g, u, dfs_count, reached, dfsnum, lowpt, parent);
+                if(lowpt[u] == dfsnum[v]){
+                    if(u == w && parent[v] != INT_MAX) {
+                        g.addEdge(u, parent[v]);
+                        g.addEdge(parent[v], u);
+                    }
+                    if(u != w){
+                        g.addEdge(u, w);
+                        g.addEdge(w, u);
+                    }
+                }
+                lowpt[v] = std::min(lowpt[v], lowpt[u]);
+            } else {
+                lowpt[v] = std::min(lowpt[v], dfsnum[u]);
+            }
+        }
+    }
+
+    void HopcroftTarjan::make_biconnected_graph(NetworKit::Graph &g) {
+        // make it connected
+        NetworKit::node u = 0;
+        std::vector<bool> reached(N);
+        for(auto v: g.nodeRange()){
+            if(!reached[v]) {
+                dfs(g, v, reached);
+                if(u != v) {
+                   g.addEdge(u, v);
+                   g.addEdge(v, u);
+                }
+            }
+        }
+        // make it biconnected
+        fill(reached.begin(), reached.end(), false);
+        std::vector<NetworKit::node> lowpt(N);
+        std::vector<NetworKit::node> parent(N, INT_MAX);
+        std::vector<NetworKit::node> dfsnum(N);
+        int dfs_count = 0;
```

Comment:
Please replace finding biconnected components with https://networkit.github.io/dev-docs/cpp_api/classNetworKit_1_1BiconnectedComponents.html#_CPPv4N9NetworKit21BiconnectedComponentsE

In general it's worthwhile to look at the NetworKit C++ docs to see if one can avoid reimplementing lower-level algorithms. E.g. simple DFS is there too and AFAIK you can pass a lambda there, so that you should be able to build `lowpt` etc.